### ISSUE TITLE
carlaviz 8081 port closed after a while in k8s env

### ISSUE THREAD

**Issue by jewes** [mjxu96/carlaviz#25]
```
Frontend launched. Please open your browser
2021-04-28 02:34:59 [INFO]     Waiting for a drawing client to connect. Listening to port 8089....
2021-04-28 02:34:59 [INFO]     Waiting for a frontend to connect. Listening to port 8081....
2021-04-28 02:38:06 [INFO]     Frontend connection closed
2021-04-28 02:38:07 [INFO]     Frontend connection closed
2021-04-28 02:38:08 [INFO]     Frontend connection closed
2021-04-28 02:38:09 [INFO]     Frontend connection closed
2021-04-28 02:38:54 [ERROR]    Connection reset by peer
```

carlaviz started, and after a while, port 8081 is closed with the above error message printed. both 0.9.10 and 0.9.11 have this issue.

**Comment by mjxu96** [comment]
It seems your frontend always closes the connection. Could you check your browser's console output? F12 to open the browser console for Chrome.

**Comment by jewes** [comment]
> It seems your frontend always closes the connection. Could you check your browser's console output? F12 to open the browser console for Chrome.

browser console does not have any output.

**Comment by jewes** [comment]
@mjxu96 i can successfully running carla and carlaviz using docker directly on a host. now I want to running them inside a k8s pod, and encountered the above issue.

**Comment by jewes** [comment]
@mjxu96 besides error message printed in the console, where can i find more log message?

**Comment by mjxu96** [comment]
CarlaViz uses 3 ports. 8080 for frontend visualization, 8081 for data transmission in websocket protocol and 8089 (optional) for drawing. Could you check you k8s firewall setting and ensure it does not block anyone of them?

**Comment by mjxu96** [comment]
> @mjxu96 besides error message printed in the console, where can i find more log message?

Hmm... Backend can only prints these logs.. Not sure why it always disconnects because in the backend's logic, I do not intentionally close the connection.

**Comment by jewes** [comment]
> CarlaViz uses 3 ports. 8080 for frontend visualization, 8081 for data transmission in websocket protocol and 8089 (optional) for drawing. Could you check you k8s firewall setting and ensure it does not block anyone of them?

Firewall does not block the 3 ports. The symptom is, I run 'netstat -lnt' command when carlaviz just started, I can see the three ports are listening. After a while, 8081 is not listening anymore (only 8080, 8089 are listening). Looks like the server process listening on 8081 crashed.
```
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN
tcp        0      0 0.0.0.0:8089            0.0.0.0:*               LISTEN
```

**Comment by jewes** [comment]
```
void FrontendProxy::Accept() {
  CARLAVIZ_LOG_INFO("Waiting for a frontend to connect. Listening to port %u....",
           frontend_listen_port_);
  try {
    boost::asio::io_context ioc{1};
    tcp::acceptor acceptor{ioc,
                           tcp::endpoint(tcp::v4(), frontend_listen_port_)};
    for (;;) {
      tcp::socket socket{ioc};
      acceptor.accept(socket);
      AddClient(std::move(socket));
    }
  } catch (const std::exception& e) {
    CARLAVIZ_LOG_ERROR("%s", e.what());
  }
}
```

Looks like something wrong when calling acceptor.accept(socket);, then break the for loop, so it no longer listen on 8081 now.

**Comment by mjxu96** [comment]
> ```
> 
> void FrontendProxy::Accept() {
> 
>   CARLAVIZ_LOG_INFO("Waiting for a frontend to connect. Listening to port %u....",
> 
>            frontend_listen_port_);
> 
>   try {
> 
>     boost::asio::io_context ioc{1};
> 
>     tcp::acceptor acceptor{ioc,
> 
>                            tcp::endpoint(tcp::v4(), frontend_listen_port_)};
> 
>     for (;;) {
> 
>       tcp::socket socket{ioc};
> 
>       acceptor.accept(socket);
> 
>       AddClient(std::move(socket));
> 
>     }
> 
>   } catch (const std::exception& e) {
> 
>     CARLAVIZ_LOG_ERROR("%s", e.what());
> 
>   }
> 
> }
> 
> ```
> 
> 
> 
> Looks like something wrong when calling acceptor.accept(socket);, then break the for loop, so it no longer listen on 8081 now.
> 
> @mjxu96 

As you can see, all exceptions are caught and printed out.

**Comment by mjxu96** [comment]
If the local network works, I assume the logic should also work in k8s. Maybe you can build the project on your own in k8s pod and print more debug information. 

I would really appreciate it if you could share what you have found.

**Comment by jewes** [comment]
> If the local network works, I assume the logic should also work in k8s. Maybe you can build the project on your own in k8s pod and print more debug information.
> 
> I would really appreciate it if you could share what you have found.

Sure, thanks!

**Comment by jewes** [comment]
After some debugging, we found the root cause. in k8s environment, the health check regularly access port 8081 to check. however, the embedded ws server in carlaviz expects valid data format which is not in the health check. so, this caused the ws server throw an exception and server crashed.

```
void FrontendProxy::AddClient(boost::asio::ip::tcp::socket socket) {
  auto client_ptr =
      boost::make_shared<websocket::stream<tcp::socket>>(std::move(socket));
  try {
    client_ptr->accept();
    CARLAVIZ_LOG_INFO("Frontend connected");

    auto frontend_client = boost::make_shared<FrontendClient>(client_ptr, 
      stream_settings_callback_);

    frontend_client->SetBinary(true);
    SendMetadata(frontend_client);

    std::lock_guard<std::mutex> lock_guard(add_client_lock_);
    frontend_clients_.insert({frontend_max_id_, frontend_client});
    frontend_max_id_++;
  } catch (boost::system::system_error const& se) {
    if (se.code() != websocket::error::closed) {
      throw se;  /// this line caused the issue.
    } else {
      CARLAVIZ_LOG_INFO("Frontend connection closed");
    }
  }
}
```

**Comment by jewes** [comment]
@mjxu96 send a pr(https://github.com/mjxu96/carlaviz/pull/27) regarding to this issue, would you please take a look? if it is ok, would you please merge it back to 0.9.10 and create a new release. We would like to use it in our project. Thanks!

**Comment by mjxu96** [comment]
@jewes Just merged your PR and triggered 0.9.11 (for Carla 0.9.11) and 0.9.10.2 (for Carla 0.9.10.1) docker image build.

**Comment by jewes** [comment]
> @jewes Just merged your PR and triggered 0.9.11 (for Carla 0.9.11) and 0.9.10.2 (for Carla 0.9.10.1) docker image build.

Great, thanks!
